---
title: "beta_diversity - USEFUL, milk samples"
author: "Filippo Biscarini"
date: "September 23, 2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
## SET UP
library("ape")
library("vegan")
library("ggplot2")
library("phyloseq")
library("tidyverse")
library("data.table")
library("metagenomeSeq")

## PARAMETERS
HOME <- Sys.getenv("HOME")
prj_folder = file.path(HOME, "Documents/USEFUL")
analysis_folder = "Analysis/microbiome_chlorine/micca"
conf_file = "Config/mapping_file_chlorine.csv"
outdir = file.path(analysis_folder)

repo = file.path(prj_folder, "metabarcoding")
source(file.path(repo, "support_functions/dist2list.R")) ## from: https://github.com/vmikk/metagMisc/
source(file.path(repo, "support_functions/phyloseq_transform.R")) ## from: https://github.com/vmikk/metagMisc/
```

```{r data, echo=FALSE}
## loading data previously imported in phyloseq
fname = file.path(prj_folder, analysis_folder, "otu_table/phyloseq.RData")
load(fname)

writeLines(" - CSS normalization")
otu_tax_sample_norm = phyloseq_transform_css(otu_tax_sample, norm = TRUE, log = FALSE)
```

```{r metadata, echo=FALSE}
## METADATA
metadata <- sample_data(otu_tax_sample_norm)
metadata$`sample-id` = row.names(metadata)
row.names(metadata) <- NULL
metadata <- as_tibble(metadata)
```

```{r beta, echo=FALSE}
## beta diversity
distances = distance(otu_tax_sample_norm, method="bray", type = "samples")
dd = dist2list(distances, tri = FALSE)
dx = spread(dd, key = "col", value = "value")

temp = dplyr::select(metadata, c(`sample-id`,type,treatment))
dx <- dx %>% inner_join(temp, by = c("row" = "sample-id"))
```

```{r mds, echo=FALSE, results='hide', message=FALSE}
# 
temp = select(dx,-row)
nvars = 2
matx= data.matrix(temp[,seq(1,ncol(temp)-nvars)])

## HULL - timepoint + treatment
udder.mds= metaMDS(matx, k = 3) #function metaMDS in Vegan

udder.scores <- as.data.frame(scores(udder.mds, display = "sites"))  #Using the scores function from vegan to extract the site scores and convert to a data.frame
udder.scores$type <- temp$type #  add the grp variable created earlier
udder.scores$treatment <- temp$treatment
```

```{r hull, echo=FALSE}
hull_f <- function(df) {
  
  temp <- data.frame(NULL)
  for (ll in unique(df$type)) {
    
    nn <- df[df$type == ll,][chull(df[df$type == ll, c("NMDS1","NMDS2","NMDS3")]),]
    temp <- rbind.data.frame(temp,nn)
  }
  return(temp)
}

hull.data <- hull_f(udder.scores)
```

## 3D plots

#### All samples

```{r 3d-plot, echo=FALSE, message=FALSE, warning=FALSE}
## 3D plot
library("plotly")

p <- plot_ly(data = udder.scores, 
             x = ~NMDS1, y = ~NMDS2, z = ~NMDS3,
             type = "scatter3d",
             color = ~treatment) %>%
  add_markers() %>%
  layout(scene = list(xaxis = list(title = 'NMDS1'),
                      yaxis = list(title = 'NMDS1'),
                      zaxis = list(title = 'NMDS1')),
         annotations = list(
           x = 0.005,
           y = 0.01,
           text = 'treatment',
           xref = 'paper',
           yref = 'paper',
           showarrow = FALSE
         ))

p
```

#### Tank stalla

```{r, echo=FALSE,  message=FALSE, warning=FALSE}
tipos = unique(udder.scores$type)
temp = filter(udder.scores, type == tipos[1])
## 3D plot
library("plotly")

p <- plot_ly(data = temp, 
             x = ~NMDS1, y = ~NMDS2, z = ~NMDS3,
             type = "scatter3d",
             color = ~treatment) %>%
  add_markers() %>%
  layout(scene = list(xaxis = list(title = 'NMDS1'),
                      yaxis = list(title = 'NMDS1'),
                      zaxis = list(title = 'NMDS1')),
         annotations = list(
           x = 0.005,
           y = 0.01,
           text = 'treatment',
           xref = 'paper',
           yref = 'paper',
           showarrow = FALSE
         ))

p
```

#### Caldaia

```{r, echo=FALSE,  message=FALSE, warning=FALSE}
temp = filter(udder.scores, type == tipos[2])
## 3D plot
library("plotly")

p <- plot_ly(data = temp, 
             x = ~NMDS1, y = ~NMDS2, z = ~NMDS3,
             type = "scatter3d",
             color = ~treatment) %>%
  add_markers() %>%
  layout(scene = list(xaxis = list(title = 'NMDS1'),
                      yaxis = list(title = 'NMDS1'),
                      zaxis = list(title = 'NMDS1')),
         annotations = list(
           x = 0.005,
           y = 0.01,
           text = 'treatment',
           xref = 'paper',
           yref = 'paper',
           showarrow = FALSE
         ))

p
```

#### Massa-caseificio

```{r, echo=FALSE, message=FALSE, warning=FALSE}
temp = filter(udder.scores, type == tipos[3])
## 3D plot
library("plotly")

p <- plot_ly(data = temp, 
             x = ~NMDS1, y = ~NMDS2, z = ~NMDS3,
             type = "scatter3d",
             color = ~treatment) %>%
  add_markers() %>%
  layout(scene = list(xaxis = list(title = 'NMDS1'),
                      yaxis = list(title = 'NMDS1'),
                      zaxis = list(title = 'NMDS1')),
         annotations = list(
           x = 0.005,
           y = 0.01,
           text = 'treatment',
           xref = 'paper',
           yref = 'paper',
           showarrow = FALSE
         ))

p
```

#### Siero-innesto

```{r, echo=FALSE, message=FALSE, warning=FALSE}
temp = filter(udder.scores, type == tipos[4])
## 3D plot
library("plotly")

p <- plot_ly(data = temp, 
             x = ~NMDS1, y = ~NMDS2, z = ~NMDS3,
             type = "scatter3d",
             color = ~treatment) %>%
  add_markers() %>%
  layout(scene = list(xaxis = list(title = 'NMDS1'),
                      yaxis = list(title = 'NMDS1'),
                      zaxis = list(title = 'NMDS1')),
         annotations = list(
           x = 0.005,
           y = 0.01,
           text = 'treatment',
           xref = 'paper',
           yref = 'paper',
           showarrow = FALSE
         ))

p
```

